### PR TITLE
Implement SeccompProfile CRD and controller

### LINKED ISSUE #117: Custom Resource Definition (CRD) for profiles
The idea is to create a CRD on top of profiles, which helps to:

- focus on a better user experience
- provides a useful abstraction, like the minimum amount of syscalls needed for container runtimes like runc
- easier create profiles on a higher level

### PR DESCRIPTION
**Pull request by cmurphy**
This change implements a SeccompProfile custom resource API. A new
SeccompProfile type is created and the controller Reconciler is modified
to handle either a SeccompProfile Kind or a ConfigMap describing a
seccomp profile. The CRD manifest is generated by the new type.

#### What type of PR is this?

/kind feature
/kind api-change

#### Which issue(s) this PR fixes:

Fixes #117 

#### Does this PR introduce a user-facing change?

```release-note
Added new Custom Resource Definition `seccompprofiles.seccomp-operator.k8s-sigs.io` as an alternative to an annotated ConfigMap for defining seccomp profiles.
```

### CODE REVIEW

File: deploy/operator.yaml
Comment (saschagrunert) [inline deploy/operator.yaml:48]: Yep, let's add this to https://github.com/kubernetes-sigs/seccomp-operator/blob/master/deploy/base/rbac.yaml and run `make deployments` (kustomize has to be installed) to sync it up.

**Review by saschagrunert — COMMENTED** [review]
Great work so far, just two nits on the deployment.

File: deploy/operator.yaml
Comment (cmurphy) [reply deploy/operator.yaml:48]: done

File: deploy/operator.yaml
Comment (cmurphy) [reply deploy/operator.yaml]: done

File: deploy/namespace-operator.yaml
Comment (hasheddan) [inline deploy/namespace-operator.yaml:44]: Do we need create permissions on the `SeccompProfile` resource?

File: deploy/base/crd.yaml
Comment (hasheddan) [inline deploy/base/crd.yaml:2]: In regards to generating, I think it would be good to do this from the beginning to make sure we stay in sync. This is fairly straightforward with `controller-tools`. Here is an example of doing so with go generate --> https://github.com/crossplane/crossplane/blob/3ea5d036778e0c257960983d0cdb9271f12dc895/apis/generate.go#L26

File: cmd/seccomp-operator/main.go
Comment (hasheddan) [inline cmd/seccomp-operator/main.go:48]: Do we need to add the `clientgoscheme` here?

File: cmd/seccomp-operator/main.go
Comment (hasheddan) [inline cmd/seccomp-operator/main.go:115]: nitpick: you can leave this empty and `ctrl.NewManager()` will create a `*runtime.Scheme` for you, which you can then add types to with something like:
```
if err := seccompoperatorv1alpha1.AddToScheme(mgr.GetScheme()); err != nil {
	return errors.Wrap(err, "Cannot add core Seccomp APIs to scheme")
}
```

**Review by hasheddan — COMMENTED** [review]
Thanks for the awesome work here @cmurphy! Excited to see this coming along! 🎉

File: deploy/namespace-operator.yaml
Comment (cmurphy) [reply deploy/namespace-operator.yaml:44]: Nope, will remove

**Comment by saschagrunert** [comment]
@cmurphy feel free to drop the `WIP` if you think this is ready for another round of review :blush:

**Comment by cmurphy** [comment]
@saschagrunert I will once I get the tests fixed :smile:

File: api/v1alpha1/groupversion_info.go
Comment (cmurphy) [inline api/v1alpha1/groupversion_info.go:19]: Need input on what the group name for this resource should be. `seccomp-operator.sigs.k8s.io` would make sense but trying to use it produces this error:

```
The CustomResourceDefinition "seccompprofiles.seccomp-operator.sigs.k8s.io" is invalid: metadata.annotations[api-approved.kubernetes.io]: Required value: protected groups must have approval annotation "api-approved.kubernetes.io", see https://github.com/kubernetes/enhancements/pull/1111
```

**Review by cmurphy — COMMENTED** [review]
This is more or less ready for feedback.

File: api/v1alpha1/groupversion_info.go
Comment (saschagrunert) [reply api/v1alpha1/groupversion_info.go:19]: I'm not sure if we would get approval if we request it from the API reviewers. :thinking:  Otherwise we don't have other options to use a pseudo name, right?

File: api/v1alpha1/seccompprofile_types.go
Comment (saschagrunert) [inline api/v1alpha1/seccompprofile_types.go:61]: Let's add a validation here too: Linux system calls can have at most 6 arguments.

File: cmd/seccomp-operator/main.go
Comment (saschagrunert) [inline cmd/seccomp-operator/main.go:119]: Let's start with lowercase errors and keep the description as short as possible to keep the noise low in the error stack:
```suggestion
		return errors.Wrap(err, "add core seccomp APIs to scheme")
```

File: internal/pkg/controllers/profile/profile.go
Comment (saschagrunert) [inline internal/pkg/controllers/profile/profile.go:120]: Is this `saver` still required?

File: go.mod
Comment (saschagrunert) [inline go.mod:15]: Do we need this? It looks like it's not imported anywhere.

**Comment by saschagrunert** [comment]
I think we should add e2e tests around this feature, too. But we can do it as follow-up PR. :)

File: internal/pkg/controllers/profile/profile.go
Comment (cmurphy) [reply internal/pkg/controllers/profile/profile.go:120]: I added it to help with the unit tests. Without it, [this new test](https://github.com/kubernetes-sigs/seccomp-operator/pull/125/files#diff-e4e0872e41f6b6c837b6e05708f07bf2R146-R154) tries to write the profile to /var/lib/. I'm new to testing and mocking in Go so open to other suggestions for this.

File: go.mod
Comment (cmurphy) [reply go.mod]: fixed

File: cmd/seccomp-operator/main.go
Comment (cmurphy) [reply cmd/seccomp-operator/main.go]: fixed

File: api/v1alpha1/seccompprofile_types.go
Comment (cmurphy) [reply api/v1alpha1/seccompprofile_types.go:62]: done

File: internal/pkg/controllers/profile/profile.go
Comment (saschagrunert) [reply internal/pkg/controllers/profile/profile.go:120]: Ah okay, yeah this seems reasonable. Another option would be to use an interface, but this function seems a good leightweight alternative to me. 👍

**Review by saschagrunert — APPROVED** [review]
/lgtm
/approve 
/hold 
PTAL @hasheddan @pjbgf

**Comment by cmurphy** [comment]
Noticed errors on the validation of flags and architectures, latest revision fixes them

**Comment by cmurphy** [comment]
Added an e2e test.

**Review by saschagrunert — APPROVED** [review]
Thank you! PTAL @hasheddan @pjbgf

File: api/v1alpha1/groupversion_info.go
Comment (pjbgf) [reply api/v1alpha1/groupversion_info.go:19]: Tricky one! Would you be opposed to use something internal (without k8s-sigs.io) and open an issue for us to resolve this later on depending on whether or not the project gets renamed? I'd rather we go for a name submission/approval only once if possible.

We probably could start with `seccompprofiles.seccomp-operator.io` and then change and submit for approval when/if we finally decide on the project renaming.

**Review by pjbgf — CHANGES_REQUESTED** [review]
Really good stuff, thank you for this contribution. 
Overall it is a lgtm from me once we change the group to a temporary one.

File: api/v1alpha1/groupversion_info.go
Comment (saschagrunert) [reply api/v1alpha1/groupversion_info.go:19]: Got a response in the [slack message where I raised the question](https://kubernetes.slack.com/archives/C5P3FE08M/p1600968242008600)

So I guess `x-k8s.io` is the namespace we have to choose.

Xref: https://github.com/kubernetes-sigs/cluster-api/issues/1073#issuecomment-514265501

Comment (cmurphy) [reply api/v1alpha1/groupversion_info.go]: Done https://github.com/kubernetes-sigs/seccomp-operator/issues/135

Comment (cmurphy) [reply api/v1alpha1/groupversion_info.go]: Fixed

**Review by pjbgf — APPROVED** [review]
/lgtm

**Review by pjbgf — APPROVED** [review]
/lgtm

**Review by hasheddan — APPROVED** [review]
/lgtm

Thanks for the awesome work here @cmurphy!!

**Comment by saschagrunert** [comment]
Great work @cmurphy!  Are there any follow ups we can consider?

**Comment by cmurphy** [comment]
@saschagrunert yes, I just propose https://github.com/kubernetes-sigs/seccomp-operator/pull/138 as an idea to address #131. I think we might also want another PR to address part of #117 that is not addressed here, "provides a useful abstraction, like the minimum amount of syscalls needed for container runtimes like runc". Also, documentation :smile:

**Pull request merged**
